src/SymmetricPositiveDefinite.jl

```diff
@@ -0,0 +1,174 @@
+using LinearAlgebra: svd, eig, eigen
+
+@doc doc"""
+    SymmetricPositiveDefinite{N} <: Manifold
+
+The manifold of symmetric positive definite matrices, i.e.
+
+```math
+\mathcal P(n) \coloneqq
+\bigl\{
+    x \in \mathbb R^{n\\times n} :
+    \xi^\mathrm{T}x\xi > 0 \text{ for all } \xi \in \mathbb R^{n}\backslash\{0\}
+\bigr\}
+```
+
+# Constructor
+
+    SymmetricPositiveDefinite(n)
+
+generates the $\mathcal P(n) \subset \mathbb R^{n\times n}$
+"""
+struct SymmetricPositiveDefinite{N} <: Manifold end
+SymmetricPositiveDefinite(n::Int) = SymmetricPositiveDefinite{n}()
+
+@doc doc"""
+    manifold_dimension(::SymmetricPositiveDefinite{N})
+
+returns the dimension of the manifold [`SymmetricPositiveDefinite`](@ref) $\mathcal P(n)$, N\in \mathbb N$, i.e.
+```math
+    \frac{n(n+1)}{2}    
+```
+"""
+@generated manifold_dimension(::SymmetricPositiveDefinite{N}) where {N} = N*(N+1)/2
+
+@doc doc"""
+
+"""
+struct LinearAffineMetric <: Metric end
+
+@doc doc"""
+
+"""
+struct LogEuclideanMetric <: Metric end
+
+distance(P::SymmetricPositiveDefinite{N},x,y) = distance(MetricManifold(P,LinearAffineMetric),x,y)
+function distance(P::MetricManifold{SymmetricPositiveDefinite{N},LinearAffineMetric},x,y)
+    s = real.( eigen( x,y ).values )
+    return any(s .<= eps() ) ? 0 : sqrt(  sum( abs.(log.(s)).^2 )  )
+end
+
+inner(P::SymmetricPositiveDefinite{N}, x, w, v) = inner(MetricManifold(P,LinearAffineMetric),x,w,v)
+function inner(P::MetricManifold{SymmetricPositiveDefinite{N},LinearAffineMetric},x,w,v)
+	svd1 = svd(x)
+	U = svd1.U
+	S = svd1.S
+	SInv = Diagonal( 1 ./ S )
+	return tr( w * U * SInv * transpose(U) * v * U * SInv * transpose(U) )
+end
+
+norm(P::SymmetricPositiveDefinite{N},x,v) = sqrt( inner(P,x,v,v) )
+norm(P::MetricManifold{SymmetricPositiveDefinite{N},LinearAffineMetric},x,v) = sqrt( inner(P,x,v,v) )
+
+function exp!(P::SymmetricPositiveDefinite{N},y,x,v) = exp!(MetricManifold(SymmetricPositiveDefinite,LinearAffineMetric),y,x,v)
+function exp!(P::MetricManifold{SymmetricPositiveDefinite{N},LinearAffineMetric}, y, x, v)
+	svd1 = svd(x)
+	U = svd1.U
+	S = svd1.S
```

Comment:
There is something wrong with indentation here.

Reply:
As above, if `x` is spd, then SVD = EIGEN. The advantage of using `eigen` is to save saving the `Vt`. If you're afraid that `x` is not symmetric, then you could wrap it with `Symmetric(x)`. This is in any case to be recommended over computing things like `T + transpose(T)` below: that has n^2 operations, and still doesn't capture symmetry in the type. Rather, linalg functions will go through the entire `x` to check for symmetry and to wrap it internally. This (i.e., the operations, the additional svd-factor, complex values) can be all avoided by wrapping as `Symmetric` what you know is symmetric. Even "worse", svd calls eigen internally https://github.com/JuliaLang/julia/pull/32017.

Reply:
I can't see whatrs wrong with the indentation, its 4 spaces; could you give a little more detail?

Reply:
@dkarrasch Concerning Symmetric I noticed that the constructor just replaces the bottom left part of the matric with the values of the right. For numerical stability that might be quite bad and building the mean might be better. Neverttheless, I changed all occurences of svd to eigen, since I really don't need `Vt`.

Reply:
It looks like a tab to me instead of four spaces. I like seeing invisible characters in editor: https://stackoverflow.com/questions/35648361/atom-editor-view-whitespace-character .

Reply:
Thanks, found it! And yes, I now also switched back to activated white space display (in VS Code though), until now VS code just did everything in 4 spaces so I never bothered ;)

Reply:
I'm not sure I follow on numerical stability. In theory, these matrices are exactly symmetric, is that right? So, your only source of asymmetry is floating point issues, which, especially for not so big matrices should be on the order of a couple machine epsilons. Now, I don't see why that error should be "symmetric about the truth", such that symmetrizing by addition would help. I'd assume that the result is as good or bad as symmetrizing by considering only the upper or only the lower half. Is there any reference that would tell that `(T + transpose(T))/2` would be any better? I'm having an especially hard time to understand this since the dependence of the eigendecomposition on the matrix entries within the class of symmetric matrices is Lipschitz. I'd expect the deviation in the eigendecomposition to be again of a handful of epsilons. Eigendecompositions of symmetric matrices are among the best-posed numerical problems you can think of.

Reply:
Well, maybe that's better when using `Symmetric`; I don't have a reference, just an expierience with my old (Matlab) code, which not necessarily used symmetry in the svd/eigen. So if you say it's stable enough, I switch to maybe doing nothing – since for an array input I do not want to return a `Symmetric` type.

Reply:
Not wanting to return a `Symmetric` is fine, I guess. Internally, I'd definitely wrap symmetric objects (I mean arguments to functions that you assume are (numerically) symmetric already) with `Symmetric`, to reduce implicit symmetry checks in the linalg functions, and to make sure `eigen` returns real eigenvalues etc. There is also some current effort on symmetrizing functions: https://github.com/JuliaLang/julia/pull/31836/files that you may find of interest.

And I think now I see where this is coming from. Of course, if you don't have a `Symmetric` wrapper and don't pass exactly symmetric matrices (that you may have gotten from previous functions), then you're in trouble, I agree. So my recommendation would be to internally wrap, and return a symmetrized matrix.

OTOH, one could think of enforcing symmetry by type, just wrap whatever comes along. After all, it's the manifold of symmetric matrices, at no point should you ever end up with something non-symmetric.

Reply:
Ah yes, for performance I already used `Symmetric` before passing to `eigen` :) and you're right, in my old code it was exactely due to imprecissions from previous calculations.

I think it might be a trail of thought to even always return symmetric; I'll think about that, but at first glance it seems reasonable.